src/protocol.rs

```diff
@@ -192,6 +196,12 @@ impl<'a> Protocol<'a> {
             }
             "p2p-websocket-star" => Ok(Protocol::P2pWebSocketStar),
             "p2p-webrtc-star" => Ok(Protocol::P2pWebRtcStar),
+            "webrtc" => Ok(Protocol::WebRTC),
+            "certhash" => {
+                let s = iter.next().ok_or(Error::InvalidProtocolString)?;
+                let decoded = multibase::Base::Base58Btc.decode(s)?;
```

Comment:
```suggestion
                let decoded = multibase::decode(s)?;
```

Given that a `/certhash` is a multibase encoded multihash, this could have any base, not just `Base58Btc`. Am I missing something?

For reference, see Golang implementation: https://github.com/multiformats/go-multiaddr/pull/176/files

Reply:
Is it a good idea to allow multiple different bases? It means that multiple multiaddresses with different string representations might actually be equal, which is rather surprising to me.

Reply:
> Is it a good idea to allow multiple different bases? It means that multiple multiaddresses with different string representations might actually be equal, and it means that decoding then re-encoding a multiaddress might produce a different result than the original, both of which are rather surprising to me.

I also don't see any immediate neither long term benefits.

Reply:
Not sure why we're having this discussion on the Rust repo. https://github.com/multiformats/multiaddr/pull/130 is the PR that added the multiaddr component, and it seems like this discussion would logically belong there.

The reason for using a multibase is the same as for any multi-thing: "We're probably wrong today, but at least we keep our way open to be right tomorrow."
If there's no need to decide for one blessed way of doing things, why would we pick one?

More specifically, different environments (e.g. browser URLs) have different encoding requirements. Multibase makes it easy to adapt to these.

Reply:
https://github.com/multiformats/multiaddr/pull/130 doesn't even define the format of the payload behind `/certhash`. There's been no discussion about this payload.

> If there's no need to decide for one blessed way of doing things, why would we pick one?

There *is* some need, as otherwise multiple multiaddresses with different string representation might be equal, as mentioned above.

Choosing which encoding to use is an arbitrary decision, but "leave the choice to the user" is a decision as well, and in my opinion worse in strictly all ways compared to arbitrarily picking one.

Reply:
_Oh, Marten raced me to it here. Slightly adjusted reply below._

> Also, I suppose that my concern regarding the fact that we're using a multihash has simply been ignored, as usual for my concerns.

Sorry in case you feel like your points are being ignored. This is definitely not deliberate.

> Not sure why we're having this discussion on the Rust repo. https://github.com/multiformats/multiaddr/pull/130 is the PR that added the multiaddr component, and it seems like this discussion would logically belong there.

Good point. In favor of moving the discussion there.

https://github.com/multiformats/multiaddr/pull/130#issuecomment-1184274939